src/exclusion/object_filter.hpp

```diff
@@ -22,35 +24,131 @@
 namespace ddwaf::exclusion {
 
 class path_trie {
-public:
-    path_trie() = default;
-
-    [[nodiscard]] path_trie find(std::string_view key) const;
-
-    template <typename T> [[nodiscard]] path_trie find(const std::vector<T> &path) const;
-
-    template <typename T> void insert(const std::vector<T> &path);
-
-    [[nodiscard]] bool is_terminal() const { return root && root->terminal; }
-    [[nodiscard]] bool is_valid() const { return root && !root->values.empty(); }
-
-protected:
+private:
     struct trie_node {
-        // Heterogenous lookups required...
-        std::unordered_map<std::string_view, std::shared_ptr<trie_node>> values;
-        // TODO this shouldn't be necessary
-        bool terminal{false};
+        std::string path_component;
+        std::vector<trie_node> children{};
+
+        explicit trie_node(std::string path_component) :
+            path_component{std::move(path_component)} {}
+
+        ~trie_node() = default;
+        trie_node(const trie_node&) = delete;
+        trie_node(trie_node&&) = default;
+        trie_node& operator=(const trie_node&) = delete;
+        trie_node& operator=(trie_node&&) = default;
+
+        struct comp {
+            bool operator()(const trie_node& n, std::string_view v) {
+                return std::string_view{n.path_component} < v;
+            }
+            bool operator()(std::string_view v, const trie_node& n) {
+                return v < std::string_view{n.path_component};
+            }
+        };
+
+        [[nodiscard]] trie_node const *get_child(std::string_view key) const {
```

Comment:
Most would probably benefit from the inlining. What about moving the whole thing inside object_filter.cpp? The only reason the type is public is because of testing, right?

Reply:
Yes, moving it entirely to the cpp would make it difficult to test independently, on the other hand it's a bit of a balance between inlining and recompilation pain as when you modify a header it results in recompiling many files.